### PR TITLE
all: fix for 1.21

### LINKED ISSUE #139: go test fails CGO_ENABLED=0 Go 1.21.0-rc2
To get this working just add the variable and check for any other Cgo related changes in runtime/cgo.
```
CGO_ENABLED=0 go test ./...
?       github.com/ebitengine/purego/cmd        [no test files]
?       github.com/ebitengine/purego/examples/libc      [no test files]
?       github.com/ebitengine/purego/examples/objc      [no test files]
?       github.com/ebitengine/purego/internal/fakecgo   [no test files]
?       github.com/ebitengine/purego/internal/strings   [no test files]
fatal error: _cgo_pthread_key_created missing

goroutine 1 [running, locked to thread]:
runtime.throw({0x1029ad194?, 0x0?})
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/panic.go:1077 +0x40 fp=0x14000046730 sp=0x14000046700 pc=0x1028cdcd0
runtime.main()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:213 +0x3fc fp=0x140000467d0 sp=0x14000046730 pc=0x1028d067c
runtime.goexit()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/asm_arm64.s:1197 +0x4 fp=0x140000467d0 sp=0x140000467d0 pc=0x102903204

goroutine 2 [force gc (idle)]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:398 +0xc8 fp=0x14000046f90 sp=0x14000046f70 pc=0x1028d0998
runtime.goparkunlock(...)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:404
runtime.forcegchelper()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:322 +0xb8 fp=0x14000046fd0 sp=0x14000046f90 pc=0x1028d07f8
runtime.goexit()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/asm_arm64.s:1197 +0x4 fp=0x14000046fd0 sp=0x14000046fd0 pc=0x102903204
created by runtime.init.6 in goroutine 1
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:310 +0x24

goroutine 18 [GC sweep wait]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:398 +0xc8 fp=0x14000042760 sp=0x14000042740 pc=0x1028d0998
runtime.goparkunlock(...)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:404
runtime.bgsweep(0x0?)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgcsweep.go:280 +0xa0 fp=0x140000427b0 sp=0x14000042760 pc=0x1028bbdc0
runtime.gcenable.func1()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgc.go:200 +0x28 fp=0x140000427d0 sp=0x140000427b0 pc=0x1028b0888
runtime.goexit()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/asm_arm64.s:1197 +0x4 fp=0x140000427d0 sp=0x140000427d0 pc=0x102903204
created by runtime.gcenable in goroutine 1
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgc.go:200 +0x6c

goroutine 19 [GC scavenge wait]:
runtime.gopark(0x1400008e000?, 0x1029d8668?, 0x1?, 0x0?, 0x140000824e0?)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:398 +0xc8 fp=0x14000042f50 sp=0x14000042f30 pc=0x1028d0998
runtime.goparkunlock(...)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:404
runtime.(*scavengerState).park(0x102aecdc0)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgcscavenge.go:425 +0x5c fp=0x14000042f80 sp=0x14000042f50 pc=0x1028b962c
runtime.bgscavenge(0x0?)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgcscavenge.go:653 +0x44 fp=0x14000042fb0 sp=0x14000042f80 pc=0x1028b9b84
runtime.gcenable.func2()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgc.go:201 +0x28 fp=0x14000042fd0 sp=0x14000042fb0 pc=0x1028b0828
runtime.goexit()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/asm_arm64.s:1197 +0x4 fp=0x14000042fd0 sp=0x14000042fd0 pc=0x102903204
created by runtime.gcenable in goroutine 1
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgc.go:201 +0xac
FAIL    github.com/ebitengine/purego    0.265s
fatal error: _cgo_pthread_key_created missing

goroutine 1 [running, locked to thread]:
runtime.throw({0x1029d4a08?, 0x0?})
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/panic.go:1077 +0x40 fp=0x14000046730 sp=0x14000046700 pc=0x1029057b0
runtime.main()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:213 +0x3fc fp=0x140000467d0 sp=0x14000046730 pc=0x10290815c
runtime.goexit()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/asm_arm64.s:1197 +0x4 fp=0x140000467d0 sp=0x140000467d0 pc=0x1029394d4

goroutine 2 [force gc (idle)]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:398 +0xc8 fp=0x14000046f90 sp=0x14000046f70 pc=0x102908478
runtime.goparkunlock(...)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:404
runtime.forcegchelper()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:322 +0xb8 fp=0x14000046fd0 sp=0x14000046f90 pc=0x1029082d8
runtime.goexit()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/asm_arm64.s:1197 +0x4 fp=0x14000046fd0 sp=0x14000046fd0 pc=0x1029394d4
created by runtime.init.6 in goroutine 1
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:310 +0x24

goroutine 3 [GC sweep wait]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:398 +0xc8 fp=0x14000047760 sp=0x14000047740 pc=0x102908478
runtime.goparkunlock(...)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:404
runtime.bgsweep(0x0?)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgcsweep.go:280 +0xa0 fp=0x140000477b0 sp=0x14000047760 pc=0x1028f38a0
runtime.gcenable.func1()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgc.go:200 +0x28 fp=0x140000477d0 sp=0x140000477b0 pc=0x1028e8368
runtime.goexit()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/asm_arm64.s:1197 +0x4 fp=0x140000477d0 sp=0x140000477d0 pc=0x1029394d4
created by runtime.gcenable in goroutine 1
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgc.go:200 +0x6c

goroutine 4 [GC scavenge wait]:
runtime.gopark(0x14000078000?, 0x1029fd770?, 0x1?, 0x0?, 0x14000003520?)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:398 +0xc8 fp=0x14000047f50 sp=0x14000047f30 pc=0x102908478
runtime.goparkunlock(...)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/proc.go:404
runtime.(*scavengerState).park(0x102b08a00)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgcscavenge.go:425 +0x5c fp=0x14000047f80 sp=0x14000047f50 pc=0x1028f110c
runtime.bgscavenge(0x0?)
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgcscavenge.go:653 +0x44 fp=0x14000047fb0 sp=0x14000047f80 pc=0x1028f1664
runtime.gcenable.func2()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgc.go:201 +0x28 fp=0x14000047fd0 sp=0x14000047fb0 pc=0x1028e8308
runtime.goexit()
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/asm_arm64.s:1197 +0x4 fp=0x14000047fd0 sp=0x14000047fd0 pc=0x1029394d4
created by runtime.gcenable in goroutine 1
        /Users/jarrettkuklis/go/go1.21rc2/src/runtime/mgc.go:201 +0xac
FAIL    github.com/ebitengine/purego/objc       0.340s
FAIL

```

### LINKED ISSUE #140: SIGSEGV when running any test/application using go 1.21.0rc2
This is on MacOS 13.5 / Intel (16" MacBook 2019). I have tried some of my ebiten applications, ebiten tests and purego tests and example. They all seem to fail here:

raff|main:purego$ go run examples/objc/main_darwin.go 
SIGSEGV: segmentation violation
PC=0x7ff800864b62 m=0 sigcode=1
signal arrived during cgo execution

goroutine 1 [syscall, locked to thread]:
runtime.cgocall(0x10cbd80, 0xc0000b2000)
	/usr/local/go/src/runtime/cgocall.go:157 +0x4b fp=0xc0000ab4b0 sp=0xc0000ab478 pc=0x1007b2b
github.com/ebitengine/purego.RegisterFunc.func1({0xc00009c3c0?, 0x2?, 0x2?})
	/Users/rsena/go/src/github.com/ebitengine/purego/func.go:212 +0xa25 fp=0xc0000ab820 sp=0xc0000ab4b0 pc=0x10cab05
reflect.callReflect(0xc00009c1b0, 0xc0000abd88, 0xc0000abc60, 0xc0000abc68)
	/usr/local/go/src/reflect/value.go:782 +0x54c fp=0xc0000abc10 sp=0xc0000ab820 pc=0x1085c0c
reflect.callReflect(0xc00009c1b0, 0xc0000abd88, 0xc0000abc60, 0xc0000abc68)
	<autogenerated>:1 +0x45 fp=0xc0000abc40 sp=0xc0000abc10 pc=0x1095a05
reflect.makeFuncStub()
	/usr/local/go/src/reflect/asm_amd64.s:47 +0x6e fp=0xc0000abd88 sp=0xc0000abc40 pc=0x1091d8e
github.com/ebitengine/purego.Dlopen({0x10fb7b0?, 0x10dafe0?}, 0x11d5348?)
	/Users/rsena/go/src/github.com/ebitengine/purego/dlfcn.go:37 +0x1f fp=0xc0000abdc0 sp=0xc0000abd88 pc=0x10c9b3f
github.com/ebitengine/purego/objc.init.0()
	/Users/rsena/go/src/github.com/ebitengine/purego/objc/objc_runtime_darwin.go:41 +0x28 fp=0xc0000abe10 sp=0xc0000abdc0 pc=0x10ce6c8
runtime.doInit1(0x11ca600)
	/usr/local/go/src/runtime/proc.go:6739 +0xd8 fp=0xc0000abf40 sp=0xc0000abe10 pc=0x10449d8
runtime.doInit(...)
	/usr/local/go/src/runtime/proc.go:6706
runtime.main()
	/usr/local/go/src/runtime/proc.go:249 +0x374 fp=0xc0000abfe0 sp=0xc0000abf40 pc=0x1037ab4
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1650 +0x1 fp=0xc0000abfe8 sp=0xc0000abfe0 pc=0x1063661

### PR DESCRIPTION
**Pull request by TotallyGamerJet**
Closes #139 
Closes #140

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -7,7 +7,7 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
-        go: ['1.18.x', '1.19.x', '1.20.x']
+        go: ['1.18.x', '1.19.x', '1.20.x', '1.21.0-rc.2']
```
Comment (hajimehoshi) [inline .github/workflows/test.yml:10]: rc.3 is the latest now.

**Comment by TotallyGamerJet** [comment]
Leaving this as a note. Go 1.21 changed the assembler on amd64 to require the NOFRAME tag. It used to be implicit for NOSPLIT functions. NOFRAME is safe to use on platforms that don't require its presence.

**Comment by hajimehoshi** [comment]
Thank you for your hard work!!

**Comment by Julio-Guerra** [comment]
Did you understand where is this breaking change coming from? I'm wondering what's new on Go's side which lead to this problem.

**Comment by TotallyGamerJet** [comment]
It's listed in the draft release notes: https://tip.golang.org/doc/go1.21#assembler

As for the changes to fakecgo that's needed bc of the changes to speed up Cgo. Also see release notes

**Pull request merged**